modular_skyrat/modules/event_vote/code/event_vote.dm

```diff
@@ -0,0 +1,357 @@
+		return
+
+	show_votes = public_outcome
+
+	// Direct chat link is good.
+	for(var/mob/iterating_user in GLOB.player_list)
+		to_chat(iterating_user, span_infoplain(span_purple("<b>EVENT: Vote started for next event! (<a href='?src=[REF(src)];open_panel=1'>Vote!</a>)</b>")))
+		SEND_SOUND(iterating_user, sound('sound/misc/bloop.ogg')) // a little boop.
+
+	for(var/client/iterating_client in GLOB.clients)
+		var/datum/action/vote_event/event_action = new
+		iterating_client.player_details.player_actions += event_action
+		event_action.Grant(iterating_client.mob)
+		generated_actions += event_action
+
+	timer_id = addtimer(CALLBACK(src, .proc/end_vote), EVENT_VOTE_TIME)
+	vote_in_progress = TRUE
+	admin_only = FALSE
+	vote_end_time = world.time + EVENT_VOTE_TIME
+
+/// Cancels a vote outright, and does not execute the event.
+/datum/controller/subsystem/events/proc/cancel_vote(mob/user)
+	if(!vote_in_progress)
+		return
+	message_admins("EVENT: [key_name_admin(user)] cancelled the current vote.")
+	reset()
+
+/// Ends the vote there and then, and executes the event.
+/datum/controller/subsystem/events/proc/end_vote()
+	if(!LAZYLEN(votes))
+		message_admins("EVENT: No votes cast, spawning random event!")
+		if(show_votes && !admin_only)
+			for(var/mob/iterating_user in GLOB.player_list)
+				to_chat(iterating_user, span_infoplain(span_purple("<b>EVENT: No votes cast, spawning random event!</b>")))
+		reset()
+		spawnEvent()
+		return
+
+	var/list/event_weighted_list = list() //We convert the list of votes into a weighted list.
+
+	for(var/iterating_event in votes)
+		event_weighted_list[iterating_event] = LAZYLEN(votes[iterating_event])
+
+	var/highest_weight = 0
+	var/list/tying_results = list() // If we have a tie, pick a random one from the tie.
+	var/datum/round_event_control/winner
+	for(var/iterating_event in event_weighted_list)
+		if(event_weighted_list[iterating_event] > highest_weight)
+			highest_weight = event_weighted_list[iterating_event]
+			winner = iterating_event
+			tying_results = list() // Clear the tying results if there's a higher winner
+		if(event_weighted_list[iterating_event] == highest_weight)
+			tying_results += iterating_event
+
+	if(LAZYLEN(tying_results) > 1) // If there's a tie, we need to pick a random one.
+		to_chat(world, "tied vote")
+		winner = pick(tying_results)
+
+	if(!winner) //If for whatever reason the algorithm breaks, we still want an event.
+		message_admins("EVENT: Vote error, spawning random event!")
+		if(show_votes && !admin_only)
+			for(var/mob/iterating_user in GLOB.player_list)
+				to_chat(iterating_user, span_infoplain(span_purple("<b>EVENT: Vote error, spawning random event!</b>")))
+		reset()
+		spawnEvent()
+		return
+
+	message_admins("EVENT: Vote ended! Winning Event: [winner.name]")
+	if(show_votes && !admin_only)
+		for(var/mob/iterating_user in GLOB.player_list)
+			to_chat(iterating_user, span_infoplain(span_purple("<b>EVENT: Vote ended! Winning Event: [winner.name]</b>")))
```

Comment:
That's duplicated many times, I'd suggest making that its own proc with a variable to be the message to send.

Reply:
true, let's fix that